e2e/kubeclient.go

```diff
@@ -0,0 +1,194 @@
+	return c.writer.Create(ctx, obj, subResource, opts...)
+}
+
+func (c *delegatingWriteSubResourceClient) Update(ctx context.Context, obj client.Object, opts ...client.SubResourceUpdateOption) error {
+	return c.writer.Update(ctx, obj, opts...)
+}
+
+func (c *delegatingWriteSubResourceClient) Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error {
+	return c.writer.Patch(ctx, obj, patch, opts...)
+}
+
+// WriterClient comprises of all mutating functions in client.Client.
+type WriterClient interface {
+	client.Writer
+	client.StatusClient
+	SubResource(subResource string) client.SubResourceWriter
+}
+
+// delegatingWriteClient delegates all mutating functions to a writer client.
+type delegatingWriteClient struct {
+	base   client.Client
+	writer WriterClient
+}
+
+func (c *delegatingWriteClient) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	return c.base.Get(ctx, key, obj, opts...)
+}
+
+func (c *delegatingWriteClient) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	return c.base.List(ctx, list, opts...)
+}
+
+func (c *delegatingWriteClient) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	return c.writer.Create(ctx, obj, opts...)
+}
+
+func (c *delegatingWriteClient) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	return c.writer.Delete(ctx, obj, opts...)
+}
+
+func (c *delegatingWriteClient) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	return c.writer.Update(ctx, obj, opts...)
+}
+
+func (c *delegatingWriteClient) Patch(ctx context.Context, obj client.Object, patch client.Patch, opts ...client.PatchOption) error {
+	return c.writer.Patch(ctx, obj, patch, opts...)
+}
+
+func (c *delegatingWriteClient) DeleteAllOf(ctx context.Context, obj client.Object, opts ...client.DeleteAllOfOption) error {
+	return c.writer.DeleteAllOf(ctx, obj, opts...)
+}
+
+func (c *delegatingWriteClient) SubResource(subResource string) client.SubResourceClient {
+	return &delegatingWriteSubResourceClient{
+		base:   c.base.SubResource(subResource),
+		writer: c.writer.SubResource(subResource),
+	}
+}
+
+func (c *delegatingWriteClient) Status() client.SubResourceWriter {
+	return c.writer.Status()
+}
+
+func (c *delegatingWriteClient) Scheme() *runtime.Scheme {
+	return c.base.Scheme()
+}
+
+func (c *delegatingWriteClient) RESTMapper() meta.RESTMapper {
+	return c.base.RESTMapper()
+}
+
+func (c *delegatingWriteClient) Base() client.Client {
+	return c.base
+}
+
+func newDelegatingWriteClient(c client.Client, w WriterClient) DelegatingClient {
+	return &delegatingWriteClient{
+		base:   c,
+		writer: w,
+	}
+}
+
+// labelWriterClient adds common labels to all objects mutated by this client.
+type labelWriterClient struct {
```

Comment:
Interesting idea to create a client that writes the labels on every write operation. It does seem a bit heavy-handed, but I think it works for test code. And it prevents the test writer from having to worry about missing any objects they create, which is nice.